src/Eshopworld.Core/Events/IEventProducer.cs

```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Eshopworld.Core.Events
+{
+    /// <summary>
+    /// Contract that provides a way to send events to the event streaming bus
+    /// </summary>
+    public interface IEventProducer: IDisposable
+    {
+        /// <summary>
+        /// Publish an event to a topic
+        /// </summary>
+        /// <typeparam name="TEvent">The type of the event that we are sending.</typeparam>
+        /// <param name="event">The event that we are sending.</param>
+        /// <param name="topicName">The name of the topic to send the event to.</param>
+        /// <param name="partition">If partitioning is supported defines the topic partition the event will be published to.</param>
+        /// <param name="cancelToken">Token that can be passed to cancel the operation.</param>
+        Task ProduceAsync<TEvent>(TEvent @event, string topicName, int? partition = null, CancellationToken cancelToken = default) where TEvent : BaseEvent;
```

Comment:
Here you let the library take care of the envelope. If there are any things you should be able to set yourself, they could be exposed here as well (e.g. identifier). But ideally, most of the envelope fields should be in full control of the library:
- created by, at
- origin service
- correlation id